woocommerce/wc-smooth-generator
# Switch Admin UI to generating in batches, using WC's BatchProcessor tool

### Changes proposed in this Pull Request:

This changes the approach to generating stuff used in the admin UI. Instead of creating a separate Action Scheduler event for _every_ product or order that should get generated, it leverages the batch methods introduced in #125 (and refined by subsequent PRs) to generate batches of items in one event. It also makes use of WooCommerce's BatchProcessor tool to handle batch management when more than one batch is needed (if you want to generate hundreds of products, for example).

To go along with this improved background generation process, this makes some improvements to the UI on the Smooth Generator page in WP Admin. It adds a progress bar while a generation job is running, which is updated periodically via WP's heartbeat system. Plus some other miscellaneous tweaks to improve the UX of the page.

Fixes #121 

### How to test the changes in this Pull Request:

1. Navigate to Tools > Smooth Generator. Click one of the buttons to generate either 10 products or 10 orders. The page should reload and now show you a progress bar in its indeterminate mode. The "Generate" buttons on the page should now be disabled, and the "Cancel" button should be enabled. It may take a minute or two before the Action Scheduler job starts running in the background, but then the bar should change to show the job is complete. Watch the music video while you wait (if you end up with Michael Jackson's Smooth Criminal video, I'm sorry it's so weird). When the job is complete, the buttons should toggle so that "Generate" is enabled again, and "Cancel" is disabled.
2. Now try doing a much larger batch so that you will actually get to see a partially filled progress bar. Try generating something like 1,000 products. It might take a minute or two once again to get started, but then you should see a partially filled progress bar. It shouldn't take too much longer after that to complete, since it will be doing batches of 100 at a time.
3. Initiate another generation job, and cancel it. Go to Tools > Scheduled Actions and make sure the event really is gone.

### Changelog entry

> Improve background generation tasks initiated through the WP Admin UI.
